### PR TITLE
Introduce the notion of JWT authenticated workers

### PR DESCRIPTION
**Pull request by doanac**

### CODE REVIEW

File: jobserv/api/worker.py [hunk 1]
```diff
@@ -39,11 +47,37 @@ def wrapper(*args, **kwargs):
...
-        if not worker.validate_api_key(parts[1]):
-            return jsendify("Incorrect API key for host", 401)
+
+        if parts[0] == "Bearer":
+            try:
+                w = worker_from_jwt(parts[1])
+            except PyJWTError as e:
+                return jsendify(str(e), 401)
+
+            if w.name != kwargs["name"]:
```
Comment (StealthyCoder) [inline jobserv/api/worker.py:59]: I am too unfamiliar with this codebase, so I just thought I'd ask. Are both the JWT token and the `kwargs['name']` user input? Then you can just always make sure they are the same even though it is a bogus name as we will create the worker then regardless. Just making sure not anyone can just create a score of workers at will.

File: jobserv/api/worker.py [hunk 2]
```diff
@@ -29,6 +31,12 @@ def _is_worker_authenticated(worker: Worker):
         parts = key.split(" ")
         if len(parts) == 2 and parts[0] == "Token":
             return worker.validate_api_key(parts[1])
+        if len(parts) == 2 and parts[0] == "Bearer":
+            try:
+                w = worker_from_jwt(parts[1])
+                return w.name == worker.name
```
Comment (mike-sul) [inline jobserv/api/worker.py:37]: So, each worker requires a dedicated JWT token?

File: jobserv/api/worker.py [hunk 3]
```diff
@@ -29,6 +31,12 @@ def _is_worker_authenticated(worker: Worker):
...
+            except PyJWTError:
+                pass
```
Comment (mike-sul) [inline jobserv/api/worker.py:39]: So, in a case of the auth failure (PyJWTError based exception or if `w.name != worker.name`) `worker_get` (HTTP GET workers/<name>/) returns 200, at least that's how I read `def worker_get(name):` implementation?

File: jobserv/api/worker.py [hunk 4]
```diff
@@ -39,11 +47,37 @@ def wrapper(*args, **kwargs):
...
+
+        if parts[0] == "Bearer":
+            try:
+                w = worker_from_jwt(parts[1])
+            except PyJWTError as e:
+                return jsendify(str(e), 401)
+
+            if w.name != kwargs["name"]:
+                # worker can only access its self
+                return jsendify("Not found", 404)
```
Comment (mike-sul) [inline jobserv/api/worker.py:61]: Shouldn't it be 401?

File: jobserv/api/worker.py [hunk 5]
```diff
@@ -39,11 +47,37 @@ def wrapper(*args, **kwargs):
...
+
+            worker = Worker.query.filter(Worker.name == w.name).first()
+            if worker is None:
+                # This looks a little nutty - constructing this object with
+                # basically "I have no idea" data. But the worker will call
+                # us with `worker_update` on its first connection which will
+                # fill these handy but not mission-cricital fields out.
+                worker = Worker(w.name, "?", 1, 1, "?", "", 1, w.allowed_tags)
```
[the remark this answers is not in the document]
Comment (mike-sul) [reply jobserv/api/worker.py:69]: It's a bit counter-intuitive, especially in the case of 
```
@blueprint.route("workers/<name>/", methods=["DELETE"])
@worker_authenticated
def worker_delete(name):
```

File: jobserv_worker.py [hunk 6]
```diff
@@ -678,6 +716,7 @@ def get_args(args=None):
         help="""Worker name to register. If none is provided, the value of
                 /etc/hostname will be used.""",
     )
+    p.add_argument("--jwt", help="A JWT to use for API authentication")
```
Comment (mike-sul) [inline jobserv_worker.py:719]: What entity does create the `jwt` token that is supposed to specified via `--jwt` token? I suppose, this is either a human or some automation script. if so, maybe it makes sense to extend `create_jwt` command so it actually creates a JWT token not just a key/cert pair.?

Comment (mike-sul) [reply jobserv_worker.py:719]: I mean something that is done in the test below, e.g.
```
key, cert = create_jwt(["org1"])
        jobserv.worker_jwt._keys.clear()

        headers = {"kid": _keyid(cert)}
        worker = {"name": "MrJWT"}
        worker["exp"] = datetime.datetime.utcnow() + datetime.timedelta(seconds=30)
        encoded = jwt.encode(worker, key, algorithm="ES256", headers=headers)
```

File: jobserv/api/worker.py (hunk 1 above)
Comment (doanac) [reply jobserv/api/worker.py:59]: You have to have a valid token to get this far. So its a token we trust. This is a check to ensure that worker-a can't steal work for worker-b. Especially in the case where worker-a is one company and worker-b is another company.

File: jobserv/api/worker.py (hunk 2 above)
Comment (doanac) [reply jobserv/api/worker.py:37]: yes. and scaling logic (or perhaps another company that wanted to privately host their own ci workers) - can create the tokens as needed: https://github.com/foundriesio/rftm/blob/main/rfds/0006/launch.py

File: jobserv/api/worker.py (hunk 3 above)
Comment (doanac) [reply jobserv/api/worker.py:39]: good point. we should probably lock this down better. workers shouldn't be able to see each other (actually regular anonymous api users shouldn't either).

File: jobserv/api/worker.py (hunk 4 above)
Comment (doanac) [reply jobserv/api/worker.py:61]: probably. i think i should do a change before this that locks down access to listing/viewing workers. then this work can be done more sanely.

File: jobserv/api/worker.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (doanac) [reply jobserv/api/worker.py:69]: i guess we could say that only API $X creates the worker. i was just sort of following our device-gateway type thinking for this. but basically - its less code this way

File: jobserv_worker.py (hunk 6 above)
Comment (doanac) [reply jobserv_worker.py:719]: token creation needs to happen inside people's custom scaling logic. this just makes it a little easier to create a keypair that logic will depend on.

File: jobserv/api/worker.py (hunk 2 above)
Comment (mike-sul) [reply jobserv/api/worker.py:37]: How to guarantee a worker name uniqueness?

File: jobserv/api/worker.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mike-sul) [reply jobserv/api/worker.py:69]: That's what I thought, kind of JITP, LGTM, could be improved/changed if it's really needed.

**Review by mike-sul — APPROVED** [review]
LGTM for the given stage of development.

File: jobserv/api/worker.py (hunk 2 above)
Comment (doanac) [reply jobserv/api/worker.py:37]: the DB will reject it and an error will get thrown.

File: jobserv/api/worker.py (hunk 1 above)
Comment (StealthyCoder) [reply jobserv/api/worker.py:59]: I understand. :+1:

**Pull request merged**